Mathlib/Tactic/Find.lean

```diff
@@ -39,96 +92,237 @@ private partial def matchHyps : List Expr → List Expr → List Expr → MetaM
+          terms := terms.push (true, t)
+        | `(find_pattern| $ss:str) => do
+          let str := Lean.TSyntax.getString ss
+          namePats := namePats.push str
+        | `(find_pattern| $i:ident) => do
+          let n := Lean.TSyntax.getId i
+          unless (← getEnv).contains n do
+            throwErrorAt i "unknown identifier '{n}'"
+          idents := idents.push n
+        | `(find_pattern| _) => do
+            throwErrorAt arg ("Cannot search for _. " ++
+              "Did you forget to put a term pattern in parentheses?")
+        | `(find_pattern| $s:term) => do
+          let t ← elabTerm' s none
+          terms := terms.push (false, t)
+        | _ => throwErrorAt args "unexpected argument to #find"
+    | _ => throwErrorAt args "unexpected argument to #find"
+    pure {idents, namePats, terms}
+
+
+open Command
+
+/--
+The `#find` command finds definitions and lemmas in various ways. One can search by: the constants
+involved in the type; a substring of the name; a subexpression of the type; or a subexpression
+located in the return type or a hypothesis specifically. All of these search methods can be
+combined in a single query, comma-separated.
+
+1. By constant:
+   ```lean
+   #find Real.sin
+   ```
+   finds all lemmas whose statement somehow mentions the sine function.
+
+2. By lemma name substring:
+   ```lean
+   #find Real.sin, "two"
+   ```
+   restricts the search above to those lemmas that have `"two"` as part of the lemma _name_.
+
+   (Currently, substring searches _must_ be combined with other kind of queries.)
+
+3. By subexpression:
+   ```lean
+   #find _ * (_ ^ _)
+   ```
+   finds all lemmas whose statements somewhere include a product where the second argument is
+   raised to some power. The pattern can also be non-linear, as in
+   ```lean
+   #find Real.sqrt ?a * Real.sqrt ?a
+   ```
+
+4. By conclusion and/or hypothesis:
+   ```lean
+   #find ⊢ tsum _ = _ * tsum _
+   ```
+   finds all lemmas where the conclusion (the subexpression to the right of all `→` and `∀`) has the
+   given shape. If the pattern has hypotheses, they are matched against the hypotheses of
+   the lemma in any order; for example,
+   ```lean
+   #find ⊢ _ < _ → tsum _ < tsum _
+   ```
+   will find `tsum_lt_tsum` even though the hypothesis `f i < g i` is not the last.
+
+5. In combination:
+   ```lean
+   #find Real.sin, "two", tsum,  _ * _, _ ^ _, ⊢ _ < _ → _
+   ```
+   will find all lemmas which mention the constants `Real.sin` and `tsum`, have `"two"` as a
+   substring of the lemma name, include a product and a power somewhere in the type, *and* have a
+   hypothesis of the form `_ < _`.
+
+If you pass more than one such search filter, `#find` will only return lemmas which match _all_ of
+them simultaneously. At least some filter must mention a concrete name, because `#find` maintains
+an index of which lemmas mention which other constants. This is also why the _first_ use of `#find`
```

Comment:
I wonder if (for later) we can just lookup patterns in the `exact?` discrimination tree, to the case without constants.

Similarly we could use the `have?` discrimination tree (currently not cached, and rather expensive) to look up hypothesis patterns.

Reply:
> I wonder if (for later) we can just lookup patterns in the exact? discrimination tree, to the case without constants.

If there are no constants, the exact discrimination tree won’t have any useful entries, either, will it? E.g. with a constant-free pattern `(_ -> _)`, what can you do (besides a linear scan)?